src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -208,30 +232,62 @@ CoreWorkerDirectTaskSubmitter::GetOrConnectLeaseClient(
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+  auto &pending_lease_request = scheduling_key_entry.pending_lease_request_;
+
+  if (pending_lease_request.first) {
     // There's already an outstanding lease request for this type of task.
     return;
   }
-  auto it = task_queues_.find(scheduling_key);
-  if (it == task_queues_.end()) {
+
+  auto &task_queue = scheduling_key_entry.task_queue_;
+  if (task_queue.empty()) {
     // We don't have any of this type of task to run.
     return;
   }
 
+  // If pipelining is enabled, check whether we really need a new worer or whether we have
+  // enough room in an existing worker's pipeline to send the new tasks
+  if (max_tasks_in_flight_per_worker_ > 1) {
+    if (scheduling_key_entry.tot_tasks_in_flight <
+        scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
+      // The pipelines to the current workers are not full yet, so we don't need more
+      // workers.
+
+      // Find a worker with a number of tasks in flight that is less than the maximum
+      // value (max_tasks_in_flight_per_worker_) and call OnWorkerIdle to send tasks to
+      // that worker
+      for (auto active_worker_addr : scheduling_key_entry.active_workers_) {
+        RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
+                  worker_to_lease_entry_.end());
+        auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
+        if (lease_entry.tasks_in_flight_ < max_tasks_in_flight_per_worker_) {
+          OnWorkerIdle(active_worker_addr, scheduling_key, false,
+                       lease_entry.assigned_resources_);
+          break;
+        }
+      }
+
+      return;
+    }
+  }
+
   auto lease_client = GetOrConnectLeaseClient(raylet_address);
-  TaskSpecification &resource_spec = it->second.front();
+  TaskSpecification &resource_spec = task_queue.front();
   TaskID task_id = resource_spec.TaskId();
   RAY_LOG(DEBUG) << "Lease requested " << task_id;
   lease_client->RequestWorkerLease(
       resource_spec, [this, scheduling_key](const Status &status,
                                             const rpc::RequestWorkerLeaseReply &reply) {
         absl::MutexLock lock(&mu_);
 
-        auto it = pending_lease_requests_.find(scheduling_key);
-        RAY_CHECK(it != pending_lease_requests_.end());
-        auto lease_client = std::move(it->second.first);
-        const auto task_id = it->second.second;
-        pending_lease_requests_.erase(it);
+        auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+        auto &pending_lease_request = scheduling_key_entry.pending_lease_request_;
+        RAY_CHECK(pending_lease_request.first);
+        auto lease_client = std::move(pending_lease_request.first);
+        const auto task_id = pending_lease_request.second;
+        pending_lease_request = std::make_pair(nullptr, TaskID::Nil());
+        RAY_CHECK(lease_client);
```

Comment:
I don't think this check is necessary (std::move should guarantee this already).